### DISCUSSION TITLE
The outliers after fit transform are different from the outliers after saving the model, loading it and using transform.

### DISCUSSION THREAD

**Discussion by maticar92** [MaartenGr/BERTopic#1833] [Q&A]
Hi everyone! 

I am currently training a model using fit_transform(), saving it as safe-tensors and then reloading it. I am then running just transform() on a subset of the first 100 documents. The issue is the topic assignments are different with these approaches. Here is an example:

fit_transform() topic assignments: 
[-1, 117, -1, 6, 60, -1, -1, -1, -1, -1, -1, -1, -1, 1, 18, 18, 1, -1, -1, -1, -1, 9, -1, -1, -1, -1, 130, -1, -1, -1, 36, -1, -1, -1, 76, -1, 74, -1, -1, 5, -1, -1, 8, -1, -1, -1, -1, -1, -1, -1, -1, 5, -1, -1, 83, -1, -1, -1, -1, 1, -1, 6, -1, -1, -1, -1, -1, -1, -1, 103, -1, -1, -1, -1, -1, -1, 87, 87, 87, -1, 1, -1, -1, -1, -1, -1, 1, 139, 2, -1, 91, 1, -1, -1, -1, -1, -1, -1, 4, -1]

transform() topic assignments:
[ 46, 117,   6,   6,  60,   1, 130, 130, 130, 130, 130, 130, 166,  130,  18,  18,   1,   1,  94, 130,   9,   9, 187,  45, 200,  10, 130, 206,   9,   0,  36,  74,  60, 200,  76, 116,  74,  74, 154, 5, 154, 154,   8,   5, 121, 209, 209, 209, 158,  -1, 169,  91, 0,   9,  83,   9, 169, 167,  38,   1, 83,   6,  76,   6, 161, 47,  47,   9, 101, 103,  18, 209,   1, 203, 203,  87,  87,  87,  83, 133,   1,  21,   9, 235,  18, 111,  62, 139, 139, 254,  91, 1,  25,  42,   9,  48,  25,   4,   4, 219]

The difference comes from the fact that a lot of the documents assigned as outliers (-1) using fit transform are assigned to a topic when not using fit transform (a few change from one topic to another topic). I have looked at the probability matrix and transform is assigning the topic to the highest probability. That would mean that the probability of the assigned topic is higher than the probability of being an outlier. 

Is this behavior expected from the fact that transform is assigning topics using cosine similarity of topic and document embeddings? How can I avoid it?

**Comment by MaartenGr** [comment]
If you save the model using `safetensors` or `pytorch`, then the underlying dimensionality reduction and clustering algorithms will be removed. Instead, cosine similarity will be used between embeddings which can give different results. Typically, you will notice that outliers are removed quite often. You can prevent this in one of two ways, use the output probabilities of `.transform` and create a threshold yourself to assign outliers are save the model with `pickle` to use the underlying clustering model.

**Comment by maticar92** [comment]
Thank you so much! I thought so, I am now trying to save the model and ran into this issue: https://github.com/MaartenGr/BERTopic/issues/514. 

I will explore some solutions as changing the python version is not feasible for me and write it there. 

I am closing this.
